system-contracts/contracts/BatchAggregator.sol

```diff
@@ -0,0 +1,354 @@
+        opt = (opt<sub?opt:sub);
+        compressedDiff = new bytes(opt+1);
+        uint8 mask = (opt<<LENGTH_BITS_OFFSET);
+        uint256 value;
+        if (transform==opt){
+            mask |= 3;
+            value = finalValue;
+        }
+        else if (add==opt){
+            mask |= 1;
+            value = finalValue-initialValue;
+        }
+        else if (sub==opt){
+            mask |= 2;
+            value = initialValue-finalValue;
+        }
+        else{
+            require(false, "Optimal operation is not transform, add or sub");
+        }
+        uint8 diffPtr = opt;
+        for(uint8 i = 0;i<opt;i+=1){
+            compressedDiff[diffPtr] = bytes1(uint8(value & BYTE_BITMASK));
+            diffPtr -= 1;
+        }
+    }
+    function returnBatchesAndClearState() external returns (bytes memory batchInfo) {
+        for (uint256 i = 0; i < chainList.length; i += 1) {
+            uint256 chainId = chainList[i];
+            uint256 numberOfInitialWrites = initialWritesSlots[chainId].length;
+            uint256 numberOfRepeatedWrites = accesedSlots[chainId].length;
+            bytes memory stateDiffs = new bytes((numberOfRepeatedWrites+numberOfInitialWrites)*STATE_DIFF_ENTRY_SIZE);
+            
+            uint256 maxEnumIndex = 0;
+            // append initial writes
+            uint256 initialWritesPtr = 0;
+            for(uint256 i = 0;i<numberOfInitialWrites*STATE_DIFF_ENTRY_SIZE;i+=STATE_DIFF_ENTRY_SIZE){
+                bytes memory stateDiff = initialWrites[chainId][initialWritesPtr];
+                assembly{
+                    mstore(add(numberOfInitialWrites, i), stateDiff)
+                }
+                initialWritesPtr += 1;
+            }
+            // append repeated writes
+            uint256 stateDiffPtr = numberOfInitialWrites*STATE_DIFF_ENTRY_SIZE;
+            for (uint256 i = 0;i<numberOfRepeatedWrites;i+=1){
+                uint64 enumIndex = accesedSlots[chainId][i];
+                bytes memory stateDiff = uncompressedWrites[chainId][enumIndex];
+                assembly{
+                    mstore(add(numberOfInitialWrites, stateDiffPtr), stateDiff)
+                }
+                maxEnumIndex = (maxEnumIndex > enumIndex?maxEnumIndex:enumIndex);
+                stateDiffPtr += STATE_DIFF_ENTRY_SIZE;
+            }
+            uint256 enumIndexSize = bytesLength(maxEnumIndex);
+            bytes memory compressedStateDiffs = new bytes((numberOfRepeatedWrites+numberOfInitialWrites)*LOOSE_COMPRESION+numberOfRepeatedWrites*enumIndexSize+numberOfInitialWrites*DERIVED_KEY_SIZE);
+            uint256 compressionPtr = 0;
+            // compress initial writes
+            uint256 compressedStateDiffSize = 0;
+            initialWritesPtr = 0;
+            for(uint256 i = 0;i<numberOfInitialWrites*STATE_DIFF_ENTRY_SIZE;i+=STATE_DIFF_ENTRY_SIZE){
+                bytes memory stateDiff = initialWrites[chainId][initialWritesPtr];
```

Comment:
just in case, does your code work fine with the same key used for both initial write and repeated write? I.e. in the first batch it is an initial write and it is a repeated write in another one